File: DataStructures/Java/HashTable.java
public class HashTable<K, V> { 
  private class HashNode
  {
    K key;
    V value;
    HashNode next;
    
    public HashNode(K key, V value)
    {
      this.key = key;
      this.value = value;
    }
  }
  
  private ArrayList<HashNode> buckets = new ArrayList<>();
  int numBuckets= 10;
  int size;
  public HashTable()
  {
    for(int i = 0; i < numBuckets; i++)
    {
      buckets.add(null);
    }
  }
  public int getSize()
  {
    return size;
  }
  
  public boolean isEmpty()
  {
    return size == 0;
  }
  private int getBucketIndex(K key)
  {
    int hashCode = key.hashCode();
    return hashCode % numBuckets;
  }
  
  public V get(K key)
  {
    /*
     * 1. get the index of the linked list starting on 'some' bucket!
     */
    int  bucketNumber = getBucketIndex(key);
    HashNode current = buckets.get(bucketNumber);
    while(current != null)
    {
      if(current.key.equals(key))
      {
        return current.value;
      }
      current = current.next;
    }
    return null;
  }
  
  public V remove (K key)
  {
    int bucketNumber = getBucketIndex(key);
    HashNode current = buckets.get(bucketNumber);
    HashNode prev = null;
    while (current != null && !current.key.equals(key)) //find place in the linked list where key is
    {
        prev = current;
        current = current.next;
    }

    if (current == null) //not in the HashTable
    {
        return null;
    }

    V retVal = current.value; //gets the return value
    if (prev != null) {
        prev.next = current.next;
    } 
    else
    {
      buckets.set(bucketNumber, current.next); //sets the first node in the linked list to the next node after current (i.e. removes current)
    }
    size--;
    return retVal;
}
  
  
  public void add(K key, V value)
  {
    int bucketNumber = getBucketIndex(key);
    HashNode current = buckets.get(bucketNumber);
    while (current != null) 
    {
        if (current.key.equals(key)) 
        {
            current.value = value;
            return;
        }
        current = current.next;
    }

    current = buckets.get(bucketNumber); //finds the first node in the arrayList buckets
    HashNode  newNode = new HashNode(key, value); //initializes the added newNode
    newNode.next = current; //sets the newNode to first node of the array list
    size++;
  }

}

File: DataStructures/Python/HashTable.py
class HashNode:
    def __init__(self, k, v):
        self.key = k
        self.value = v
        self.next = None

class HashTable:
    def __init__(self):
        self.numBuckets = 10
        self.size = 0
        self.buckets = [None for i in range(self.numBuckets)]

    def getSize(self):
        return self.size

    def isEmpty(self):
        return self.size == 0

    def getHashCode(self, key):
        hashCode = hash(key)
        hashIndex = hashCode % self.numBuckets
        return hashIndex

    def get(self, key):
        bucketNumber = self.getHashCode(key)
        currentNode = self.buckets[bucktNumber]
        while currentNode != None:
            if currentNode.key == key:
                return currentNode.value
            currentNode = currentNode.next
        return None

    def insert(self, key, value):
        bucketNumber = self.getHashCode(key)
        currentNode = self.buckets[bucketNumber]
        newNode = HashNode(key, value)
        if currentNode == None:
            self.buckets[bucketNumber] = newNode
            self.size+= 1
        else:
            while currentNode != None:
                if(currentNode.key == key):
                    currentNode.value = value;
                    size += 1
                    break
                currentNode = currentNode.next
            if currentNode == None:
                currentNode = self.buckets[bucketNumber]
                newNode.next = currentNode
                self.buckets[bucketNumber] = newNode
                self.size+=1

    def remove(self, key):
        bucketNumber = self.getHashCode(key)
        currentNode = self.buckets[bucketNumber]
        previousNode = None
        while currentNode != None and currentNode.key != key:
            previousNode = currentNode
            currentNode = currentNode.next
        retVal = currentNode.value
        if previousNode != None:
            previousNode.next = currentNode.next
        else:
            self.buckets[bucketNumber] = currentNode.next
            self.size -= 1
        return retVal            
